### ISSUE TITLE
Map randomly fails to load

### ISSUE THREAD

**Issue by Rudloff** [JeroenDeDauw/Maps#321]
### Setup

- MW version: 1.28.1
- DB (MySQL etc.): 5.6.36
- PHP version: 5.6.30-0+deb8u1
- Maps version: 4.1.0 + 4.2.0 + 4.4.0
- SMW version (if applicable): 2.4.6

### Issue

Sometimes the map fails to load correctly and I get this error in the console:
```
19:08:49,673 Exception in module-execute in module ext.maps.leaflet: 1 load.php:176:524
19:08:49,673 ReferenceError: L is not defined ReferenceError: L is not defined
Trace de la pile :
@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=ext.maps.common%2Cleaflet&skin=archiwiki&version=1olmep4:8:1
execute/runScript/<@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=archiwiki&version=0ewpln0:161:722
jQuery.Callbacks/fire@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=archiwiki&version=0ewpln0:45:104
jQuery.Callbacks/self.add@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=archiwiki&version=0ewpln0:45:656
.Deferred/promise.always@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=archiwiki&version=0ewpln0:46:865
execute/runScript@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=archiwiki&version=0ewpln0:161:588
execute/</checkCssHandles@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=archiwiki&version=0ewpln0:162:315
execute@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=archiwiki&version=0ewpln0:163:25
mw.loader.implement@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=archiwiki&version=0ewpln0:169:267
@http://localhost/archi-mediawiki/load.php?debug=false&lang=fr&modules=ext.maps.common%2Cleaflet&skin=archiwiki&version=1olmep4:1:342
 1 load.php:176:562
	logError http://localhost/archi-mediawiki/load.php:176:562
	handler http://localhost/archi-mediawiki/load.php:155:28
	jQuery.Callbacks/fire http://localhost/archi-mediawiki/load.php:45:104
	jQuery.Callbacks/self.fireWith http://localhost/archi-mediawiki/load.php:46:431
	jQuery.Callbacks/self.fire http://localhost/archi-mediawiki/load.php:46:474
	mw.track http://localhost/archi-mediawiki/load.php:154:806
	execute/runScript/< http://localhost/archi-mediawiki/load.php:161:934
	jQuery.Callbacks/fire http://localhost/archi-mediawiki/load.php:45:104
	jQuery.Callbacks/self.add http://localhost/archi-mediawiki/load.php:45:656
	.Deferred/promise.always http://localhost/archi-mediawiki/load.php:46:865
	execute/runScript http://localhost/archi-mediawiki/load.php:161:588
	execute/</checkCssHandles http://localhost/archi-mediawiki/load.php:162:315
	execute http://localhost/archi-mediawiki/load.php:163:25
	mw.loader.implement http://localhost/archi-mediawiki/load.php:169:267
	<anonyme> http://localhost/archi-mediawiki/load.php:1:342
```

It seems to be rather random and often reloading the page will fix it.
Maybe the JS is not loaded in the right order?

**Comment by JeroenDeDauw** [comment]
@PeterTheOne any idea?

Might be fixed by https://github.com/JeroenDeDauw/Maps/issues/249

**Comment by Rudloff** [comment]
I tried updating to 4.2.0 but it still happens.

**Comment by JeroenDeDauw** [comment]
I have no idea about what the issue here is or how it can be fixed. I'm not up to speed with JS stuff in MediaWiki. You could try asking on the wikitech-l mailinglist.

**Comment by JeroenDeDauw** [comment]
I also confirm this is still happening with 4.2.0.

**Comment by Rudloff** [comment]
I can still reproduce with Maps 4.4.0 and it's pretty strange: sometimes reloading the page fixes it, sometimes I have to use `action=purge`.

**Comment by JeroenDeDauw** [comment]
This happens only for Leaflet right?

**Comment by Rudloff** [comment]
I'm not sure since we only use Leaflet and it is too random to try to reproduce with another provider.
But since the error is about Leaflet not correctly loading, I would assume it doesn't happen with other libraries.

**Comment by PeterTheOne** [comment]
Strange..

> Maybe the JS is not loaded in the right order?

Would be my best guess.. the dependencies are specified in https://github.com/JeroenDeDauw/Maps/blob/master/includes/services/Leaflet/Leaflet.php which looks good to me as far as I understand it.

**Comment by PeterTheOne** [comment]
or not.. `ext.maps.leaflet.fullscreen`, `ext.maps.leaflet.markercluster`, `ext.maps.leaflet.providers` should use `'dependencies' => [ 'ext.maps.leaflet' ],`. Could you try if this resolves the issue?

**Comment by JeroenDeDauw** [comment]
https://github.com/JeroenDeDauw/Maps/pull/374

**Comment by JeroenDeDauw** [comment]
I'm going to close this as it is presumably fixed. If not, please reopoen.

**Comment by Rudloff** [comment]
Sorry, I'm on holiday. I will try this next week.

**Comment by hidrarga** [comment]
It just happened to me recently. When I log in and I am redirected to a page with a map on it, the map fails to load. I get the exact same error.

**Comment by JeroenDeDauw** [comment]
The latest release of Maps still has this bug. You need to get the development version to get the fix for now. `5.0@dev` as version in `composer.json` + `composer update` will do that. The actual release will happen soonish.

**Comment by hidrarga** [comment]
Indeed, I don't have this bug anymore. I thought I was on the latest (dev) commit when I saw that bug, but probably not, then 😅

**Comment by Rudloff** [comment]
Sorry to come back to this bug so long after it has been closed, but we still frequently get reports from our users that the map randomly does not load.

I tried to reproduce the issue and I can still get this error (usually on the first time I load a page):
```
ReferenceError: L is not defined ReferenceError: L is not defined
Stack trace:
@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=ext.maps.common%2Cleaflet&skin=vector&version=0ow4253:8:1
runScript/<@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=vector&version=18wth5n:161:89
fire@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=vector&version=18wth5n:45:104
add@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=vector&version=18wth5n:45:656
always@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=vector&version=18wth5n:46:865
runScript@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=vector&version=18wth5n:160:944
checkCssHandles@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=vector&version=18wth5n:161:678
execute@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=vector&version=18wth5n:162:384
implement@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=jquery%2Cmediawiki&only=scripts&skin=vector&version=18wth5n:168:670
@http://www.archi-wiki.org/load.php?debug=false&lang=fr&modules=ext.maps.common%2Cleaflet&skin=vector&version=0ow4253:1:342
 load.php:176:53
```

If I remember correctly, this was fixed in 5.0.0.
The wiki I am testing on still uses 5.0.2 (the server is still Debian 7.11/PHP 5.6). I will try to get them to update to a more recent version.

As a workaround, I will try to force load Leaflet on every page. It is not very clean but it should do the job.
